File: src/domain/entity/snippets.ts
// ============================================================
// 文档片段提取（领域规则，纯函数，零基础设施依赖）
// 从 chat/eval route 的重复实现中抽出（DRY）：
//   - estimateTokens：中英文混合 token 估算
//   - mergeRanges：重叠区间合并
//   - extractEntitySnippets：长文档实体上下文片段提取（密度排序 top-N）
// ============================================================

/**
 * 估算文本的 token 数量（混合中英文场景）
 * 中文约 1.5 字符/token，英文/数字约 4 字符/token
 */
export function estimateTokens(text: string): number {
  let chineseChars = 0;
  let otherChars = 0;
  for (const ch of text) {
    // Unicode 范围：CJK 统一表意文字 + 中文标点
    if (/[\u4e00-\u9fff\u3400-\u4dbf\uff00-\uffef\u3000-\u303f]/.test(ch)) {
      chineseChars++;
    } else {
      otherChars++;
    }
  }
  return Math.ceil(chineseChars / 1.5 + otherChars / 4);
}

/**
 * 合并重叠或相邻的字符区间
 */
export function mergeRanges(ranges: Array<{ start: number; end: number }>): Array<{ start: number; end: number }> {
  if (ranges.length === 0) return [];
  const sorted = [...ranges].sort((a, b) => a.start - b.start);
  const merged: Array<{ start: number; end: number }> = [sorted[0]];

  for (let i = 1; i < sorted.length; i++) {
    const last = merged[merged.length - 1];
    if (sorted[i].start <= last.end) {
      // 有重叠，合并
      last.end = Math.max(last.end, sorted[i].end);
    } else {
      merged.push({ ...sorted[i] });
    }
  }
  return merged;
}

/**
 * 从长文档中提取实体关键字上下文的片段
 *
 * 算法：
 * 1. 用所有 matchedKeywords 构建正则，匹配文档中所有提及位置
 * 2. 对每处提及，向前后扩展指定 token 数对应的字符
 * 3. 合并重叠的区间
 * 4. 按提及密度排序，取 top-N 个片段
 */
export function extractEntitySnippets(
  content: string,
  docName: string,
  keywords: string[],
  contextTokens: number,
  maxSnippets: number,
  cnCharsPerToken: number,
  enCharsPerToken: number,
): string | null {
  // 构建正则：匹配所有关键字（按长度降序，避免短关键字被长关键字遮蔽）
  const sortedKeywords = [...keywords].sort((a, b) => b.length - a.length);
  const escaped = sortedKeywords.map((k) => k.replace(/[.*+?^${}()|[\]\\]/g, '\\$&'));
  const pattern = new RegExp(escaped.join('|'), 'gi');

  // 找到所有匹配位置（字符索引）
  const matches: Array<{ start: number; end: number; keyword: string }> = [];
  let match: RegExpExecArray | null;
  // 重置 lastIndex
  pattern.lastIndex = 0;
  while ((match = pattern.exec(content)) !== null) {
    matches.push({ start: match.index, end: match.index + match[0].length, keyword: match[0] });
  }

  if (matches.length === 0) return null;

  // 估算上下文窗口对应的字符数
  // 取中文和英文 token→字符 的平均值做近似估算
  const avgCharsPerToken = (cnCharsPerToken + enCharsPerToken) / 2;
  const contextChars = Math.ceil(contextTokens * avgCharsPerToken);

  // 对每个匹配位置扩展上下文窗口
  const rawRanges: Array<{ start: number; end: number }> = matches.map((m) => ({
    start: Math.max(0, m.start - contextChars),
    end: Math.min(content.length, m.end + contextChars),
  }));

  // 合并重叠区间
  const mergedRanges = mergeRanges(rawRanges);

  // 计算每个区间的"提及密度"（提及次数 / 区间字符数），密度高的优先
  const scoredRanges = mergedRanges.map((range) => {
    const mentionCount = matches.filter(
      (m) => m.start >= range.start && m.end <= range.end,
    ).length;
    const rangeLength = range.end - range.start;
    const density = mentionCount / (rangeLength || 1);
    return { ...range, mentionCount, density };
  });

  // 按密度降序排序，取 top-N
  scoredRanges.sort((a, b) => b.density - a.density);
  const topRanges = scoredRanges.slice(0, maxSnippets);

  // 按文档中的原始位置排序输出
  topRanges.sort((a, b) => a.start - b.start);

  // 构建输出片段
  const snippetParts = topRanges.map((range, idx) => {
    const snippet = content.slice(range.start, range.end).trim();
    const startToken = estimateTokens(content.slice(0, range.start));
    const endToken = startToken + estimateTokens(snippet);
    const header =
      topRanges.length > 1
        ? `#### 片段 ${idx + 1}（约第 ${startToken}-${endToken} token，提及 ${range.mentionCount} 次）`
        : `#### 实体上下文片段（约第 ${startToken}-${endToken} token，提及 ${range.mentionCount} 次）`;
    return `${header}\n\n${snippet}`;
  });

  return `### ${docName}（长文档片段提取，原文档约 ${estimateTokens(content)} token）\n\n${snippetParts.join('\n\n')}`;
}

File: python-server/server/rag/retrieve/snippets.py
"""文档片段提取（领域纯函数，移植自 src/domain/entity/snippets.ts）。

- estimate_tokens：中英文混合 token 估算
- merge_ranges：重叠区间合并
- extract_entity_snippets：长文档实体上下文片段提取（密度排序 top-N）
"""

from __future__ import annotations

import math
import re
from dataclasses import dataclass

_CJK_PATTERN = re.compile(r"[\u4e00-\u9fff\u3400-\u4dbf\uff00-\uffef\u3000-\u303f]")
_REGEX_SPECIAL = re.compile(r"[.*+?^${}()|[\]\\]")


@dataclass
class Range:
    start: int
    end: int


def estimate_tokens(text: str) -> int:
    """中文约 1.5 字符/token，英文/数字约 4 字符/token。"""
    chinese_chars = 0
    other_chars = 0
    for ch in text:
        if _CJK_PATTERN.match(ch):
            chinese_chars += 1
        else:
            other_chars += 1
    return math.ceil(chinese_chars / 1.5 + other_chars / 4)


def merge_ranges(ranges: list[Range]) -> list[Range]:
    if not ranges:
        return []
    sorted_ranges = sorted(ranges, key=lambda r: r.start)
    merged = [Range(sorted_ranges[0].start, sorted_ranges[0].end)]

    for r in sorted_ranges[1:]:
        last = merged[-1]
        if r.start <= last.end:
            last.end = max(last.end, r.end)
        else:
            merged.append(Range(r.start, r.end))
    return merged


def _escape_regex(word: str) -> str:
    return _REGEX_SPECIAL.sub(r"\\\g<0>", word)


def extract_entity_snippets(
    content: str,
    doc_name: str,
    keywords: list[str],
    context_tokens: int,
    max_snippets: int,
    cn_chars_per_token: float,
    en_chars_per_token: float,
) -> str | None:
    sorted_keywords = sorted(keywords, key=lambda k: -len(k))
    escaped = [_escape_regex(k) for k in sorted_keywords]
    pattern = re.compile("|".join(escaped), re.IGNORECASE)

    matches = [(m.start(), m.end()) for m in pattern.finditer(content)]
    if not matches:
        return None

    avg_chars_per_token = (cn_chars_per_token + en_chars_per_token) / 2
    context_chars = math.ceil(context_tokens * avg_chars_per_token)

    raw_ranges = [
        Range(max(0, start - context_chars), min(len(content), end + context_chars))
        for start, end in matches
    ]
    merged_ranges = merge_ranges(raw_ranges)

    scored = []
    for rng in merged_ranges:
        mention_count = sum(
            1 for start, end in matches if start >= rng.start and end <= rng.end
        )
        range_length = rng.end - rng.start
        density = mention_count / (range_length or 1)
        scored.append((rng.start, rng.end, mention_count, density))

    # 密度降序取 top-N，再按原始位置排序输出
    scored.sort(key=lambda x: -x[3])
    top = sorted(scored[:max_snippets], key=lambda x: x[0])

    parts = []
    multi = len(top) > 1
    for idx, (start, end, mention_count, _density) in enumerate(top):
        snippet = content[start:end].strip()
        start_token = estimate_tokens(content[:start])
        end_token = start_token + estimate_tokens(snippet)
        if multi:
            header = (
                f"#### 片段 {idx + 1}（约第 {start_token}-{end_token} token，"
                f"提及 {mention_count} 次）"
            )
        else:
            header = (
                f"#### 实体上下文片段（约第 {start_token}-{end_token} token，"
                f"提及 {mention_count} 次）"
            )
        parts.append(f"{header}\n\n{snippet}")

    return (
        f"### {doc_name}（长文档片段提取，原文档约 {estimate_tokens(content)} token）\n\n"
        + "\n\n".join(parts)
    )
